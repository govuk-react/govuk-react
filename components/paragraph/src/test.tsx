import * as React from 'react';
import { MemoryRouter, Link } from 'react-router-dom';
import { mount } from 'enzyme';

import { Paragraph } from '.';

const exampleCodeBlock = `\`\`\`
  const CustomStyledButton = styled(Button)({
    backgroundColor: WHITE,
    borderStyle: 'solid',
    borderWidth: '2px',
    borderColor: BLUE,
    color: BLUE,
    ':hover': {
      backgroundColor: BLUE,
      color: WHITE
    }
  });
  \`\`\``;

const ReactRouterLinkRenderer: React.FC<ReactRouterLinkRendererProps> = ({
  href,
  children,
}: ReactRouterLinkRendererProps) =>
  href.match(/^\//) ? <Link to={href}>{children}</Link> : <a href={href}>{children}</a>;

interface ReactRouterLinkRendererProps {
  href: string;
  children: React.ReactNode;
}

describe('Paragraph', () => {
  const examplePlain = 'Some basic text';
  const exampleInlineCode = '`Some inline code example`';
  const exampleEmphasis = '*Some emphasis example*';
  const exampleStrong = '**Some strong/bold example**';
  const exampleInternalLink = '[link text](/some_link)';
  const exampleExternalLink = '[link text](http://google.com)';
  let wrapper;

  it('renders a paragraph element', () => {
    wrapper = mount(<Paragraph>{examplePlain}</Paragraph>);
    expect(wrapper.find('p')).toHaveLength(1);
  });

  it('renders a inline code block', () => {
    wrapper = mount(<Paragraph>{exampleInlineCode}</Paragraph>);
    expect(wrapper.find('code')).toHaveLength(1);
  });

  it('renders a code block', () => {
    wrapper = mount(<Paragraph>{exampleCodeBlock}</Paragraph>);
    expect(wrapper.find('pre')).toHaveLength(1);
  });

  it('renders emphasis text', () => {
    wrapper = mount(<Paragraph>{exampleEmphasis}</Paragraph>);
    expect(wrapper.find('em')).toHaveLength(1);
  });

  it('renders strong text', () => {
    wrapper = mount(<Paragraph>{exampleStrong}</Paragraph>);
    expect(wrapper.find('strong')).toHaveLength(1);
  });

  it('renders react router links with custom renderer', () => {
    wrapper = mount(
      <MemoryRouter>
        <Paragraph linkRenderer={ReactRouterLinkRenderer}>{exampleInternalLink}</Paragraph>
      </MemoryRouter>
    );
    expect(wrapper.find('a')).toHaveLength(1);
  });

  it('renders internal links as anchors', () => {
    wrapper = mount(
      <MemoryRouter>
        <Paragraph>{exampleInternalLink}</Paragraph>
      </MemoryRouter>
    );
    expect(wrapper.find('a')).toHaveLength(1);
  });

  it('renders external links as anchors', () => {
    wrapper = mount(<Paragraph>{exampleExternalLink}</Paragraph>);
    expect(wrapper.find('a')).toHaveLength(1);
  });

  it('renders full example text from fixtures', () => {
    wrapper = mount(<Paragraph>{examplePlain}</Paragraph>);
  });

  it('renders as supporting text', () => {
    wrapper = mount(<Paragraph supportingText>{examplePlain}</Paragraph>);
    expect(wrapper.prop('supportingText')).toBe(true);
  });

  it('matches wrapper snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
