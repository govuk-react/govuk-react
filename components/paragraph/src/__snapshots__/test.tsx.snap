// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Paragraph matches wrapper snapshot 1`] = `
<Paragraph
  linkRenderer={[Function]}
  supportingText={true}
>
  <src__StyledParagraph
    allowedTypes={
      Array [
        "paragraph",
        "emphasis",
        "strong",
        "link",
        "inlineCode",
        "code",
        "text",
      ]
    }
    escapeHtml={false}
    linkRenderer={[Function]}
    renderers={
      Object {
        "link": [Function],
      }
    }
    skipHtml={true}
    source="Some basic text"
    supportingText={true}
  >
    <ReactMarkdown
      allowedTypes={
        Array [
          "paragraph",
          "emphasis",
          "strong",
          "link",
          "inlineCode",
          "code",
          "text",
        ]
      }
      className="src__StyledParagraph-sc-sgc9po-0 crTLg"
      escapeHtml={false}
      linkRenderer={[Function]}
      renderers={
        Object {
          "link": [Function],
        }
      }
      skipHtml={true}
      source="Some basic text"
      supportingText={true}
      transformLinkUri={[Function]}
    >
      <Root
        className="src__StyledParagraph-sc-sgc9po-0 crTLg"
        key="root-1-1-0"
        node={
          Object {
            "children": Array [
              Object {
                "children": Array [
                  Object {
                    "position": Object {
                      "end": Object {
                        "column": 16,
                        "line": 1,
                        "offset": 15,
                      },
                      "start": Object {
                        "column": 1,
                        "line": 1,
                        "offset": 0,
                      },
                    },
                    "type": "text",
                    "value": "Some basic text",
                  },
                ],
                "position": Object {
                  "end": Object {
                    "column": 16,
                    "line": 1,
                    "offset": 15,
                  },
                  "start": Object {
                    "column": 1,
                    "line": 1,
                    "offset": 0,
                  },
                },
                "type": "paragraph",
              },
            ],
            "position": Object {
              "end": Object {
                "column": 16,
                "line": 1,
                "offset": 15,
              },
              "start": Object {
                "column": 1,
                "line": 1,
                "offset": 0,
              },
            },
            "type": "root",
          }
        }
      >
        <div
          className="src__StyledParagraph-sc-sgc9po-0 crTLg"
        >
          <p
            key="paragraph-1-1-0"
          >
            <TextRenderer
              key="text-1-1-0"
              node={
                Object {
                  "position": Object {
                    "end": Object {
                      "column": 16,
                      "line": 1,
                      "offset": 15,
                    },
                    "start": Object {
                      "column": 1,
                      "line": 1,
                      "offset": 0,
                    },
                  },
                  "type": "text",
                  "value": "Some basic text",
                }
              }
              nodeKey="text-1-1-0"
              value="Some basic text"
            >
              Some basic text
            </TextRenderer>
          </p>
        </div>
      </Root>
    </ReactMarkdown>
  </src__StyledParagraph>
</Paragraph>
`;
